saltgui/static/scripts/api.js

```diff
@@ -153,40 +164,190 @@ class API {
     return this._callMethod("GET", "/jobs/" + id, {});
   }
 
-  runFunction(target, toRun) {
-    var args = toRun.split(" ");
-    var functionToRun = args[0];
-    args.shift();
+  _showError(errorMessage) {
+    var errLabel = document.querySelector("#cmd_error");
+    errLabel.innerText = errorMessage;
+    if(errorMessage)
+      errLabel.style.display = "block";
+    else
+      errLabel.style.display = "none";
+  }
+
+  _parseCommandLine(toRun, args, params) {
+    // just in case the user typed some extra whitespace
+    // at the start of the line
+    toRun = toRun.trim();
+
+    while(toRun.length > 0)
+    {
+      var name = null;
+
+      var firstSpaceChar = toRun.indexOf(" ");
+      if(firstSpaceChar < 0)
+        firstSpaceChar = toRun.length;
+      var firstEqualSign = toRun.indexOf("=");
+      if(firstEqualSign >= 0 && firstEqualSign < firstSpaceChar) {
+        // we have the name of a named parameter
+        name = toRun.substr(0, firstEqualSign);
+        toRun = toRun.substr(firstEqualSign + 1);
+        if(toRun === "" || toRun[0] === " ") {
+          this._showError("Must have value for named parameter '" + name + "'");
+          return false;
+        }
+      }
+
+      // Determine whether the JSON string starts with a known
+      // character for a JSON type
+      var endChar = undefined;
+      var objType = undefined;
+      if(toRun[0] === '{') {
+        endChar = '}';
+        objType = "dictionary";
+      } else if(toRun[0] === '[') {
+        endChar = ']';
+        objType = "array";
+      } else if(toRun[0] === '"') {
+        endChar = '"';
+        objType = "double-quoted-string";
+      } else if(toRun[0] === '\'') {
+        endChar = '\'';
+        objType = "single-quoted-string";
+      }
+
+      var value;
+      if(endChar && objType) {
+        // The string starts with a character for a known JSON type
+        var p = 1;
+        while(true) {
```

Comment:
Isn't there something you can use instead of `while (true)` ? When you make a mistake you will have an infinitive loop.

Reply:
The loop only repeats from the `continue` in error handler `catch(err)`.
All other paths have a `return` or a `break`.
The advancement of the position-counter `p` in that exception handler makes sure that the loop always ends (a infinite strings do not exist).